### ISSUE TITLE
SampleGen: Java CLI failure

### ISSUE THREAD

**Issue by hzyi-google** [googleapis/gapic-generator#2580]
For Java standalone samples, checked exception thrown by `CommandLineParser` is not caught or declared to throw, making the samples uncompilable.

- short-term solution
  - Add `throws Exception` to main method if we need to parse CLIs in a sample.
- long-term solution
  - Do not use `commons-cli`. Manually parse CLIs. This will drop the dependency of `commons-cli` altogether and avoid potential diamond dependency problems.

**Comment by hzyi-google** [comment]
cc/ @vchudnov-g

**Comment by vchudnov-g** [comment]
I agree with the short-term solution.

Long-term solution: are the diamond dependency problems an issue here with samples? I would think not, since the samples are meant to be standalone-runnable. I think we are better off relying on a standard library for CLI parsing rather than doing it ourselves, which is more work for us and bloat/distraction in the samples.

**Comment by hzyi-google** [comment]
@vam-google may have a stronger opinion on this. It _might_ be a problem for samples in the sense that the dependencies of samples might depend on a different version of `common-cli`. And since we need to make the samples standalone-runnable we need to pull in all the dependencies of dependencies, which may introduce a risk of diamond dependency there.

However, while this might be a potential problem for us, it wouldn't cause any problem for others in the near feature as no one should depend on the samples. So IMO we can push the problem back until we really have to solve this.

Also, I wouldn't expect manually parsing the CLIs to be too hard since we just need to parse `--flags` with default values, i.e., we don't need any other fancy features `common-cli` supports.

**Comment by vchudnov-g** [comment]
Yeah, parsing ourselves isn't the end of the world, but it could be extra cognitive load for users and friction for us in the future if we do need fancier features or if we need to tweak something.

I wonder whether dependencies of samples could ever depend on `common-cli`....

Curious to hear @vam-google 's thoughts. I'm OK with staying the course for now and revisiting if it does indeed become a problem or imminent problem. If we're all in agreement, maybe we can drop this to P3 and then bump it back if/when we determine we need to deal with this.

**Comment by vam-google** [comment]
@vchudnov-g, @hzyi-google  Using `common-cli` would not necessarily break something (see "Samples specifics" section below for details), I was just trying to emphasize the importance of dependencies and potential harm of extra dependencies. 

**Some context**
For `commons-cli` case specifically, we already had it in our generated SmokeTests, but then removed them (it was not my idea to remove them). The reason to do so was that this would make `google-cloud-java` depend on commons, and adding the dependency only to parse command line arguments for a smoke test was too much. 

**For libraries extra dependencies are harmful**
There is a big difference between writing an application and writing a library with regard to dependencies. For libraries dependencies are essentially part of our "surface" (for an application they are "whatever" as long as the app works). We can't hide dependencies, we require them, they put restrictions on any program, which uses our library. They make our library "heavier", less compatible with anything else, harder to integrate into an application. For example guava does not depend on anything else for a reason. There is a concept of shading dependencies, used in many places (for example grpc shades netty). Shading is painful, makes the library heavier, but people still do just because of how harmful extra dependencies are for any library. 

The problems with dependencies are especially true for very popular libraries (like guava or apache commons), because of how popular they are (so there is a good chance that there will be a conflict).

**Samples specifics**
For samples, using commons (or anything else) may be not a problem at all as long they are generated as a completely separate module (they still can be generated by same generator, what matters is the packaging of the code here). In that case samples are not library, they are an actual application. They are not designed to be integrated into anything, they are designed to be exectued as a standalone application, and thus, their dependencies are not part of the surface, they are "whatever", as long as they help to make the sample better (but we still should be careful here with regard to convergence problem).

This is different from the smoke tests (see above), because they are a part of the generated GAPIC client library.

In other words, I believe it is important to ensure that samples are generated in a separate module, so gapic library can be easily compiled independently from the samples, and samples are structured in a form of a standalone application, which also can be compiled separately from the actual gapic library it depends on. As long as this "requirement" is satisfied (I don't know if it is) the dependencies management of the samples can be much less restrictive. On another hand, probably the more portable an actual sample is the better it is for it usability (so a client can copy-paste the sample anywhere and it just works, without need to add extra dependencies).

**Comment by hzyi-google** [comment]
> I believe it is important to ensure that samples are generated in a separate module
This is not the current status but is what we want to achieve.

Thanks for the clarification @vam-google . I think we can close this issue now.
